src/index.ts

```diff
@@ -0,0 +1,465 @@
+ * @extends {Map}
+ */
+class Collection<K, V> extends Map<K, V> {
+	private _array!: V[] | null;
+	private _keyArray!: K[] | null;
+	public static readonly default: typeof Collection = Collection;
+
+	public constructor(entries?: ReadonlyArray<readonly [K, V]> | null) {
+		super(entries);
+
+		/**
+		 * Cached array for the `array()` method - will be reset to `null` whenever `set()` or `delete()` are called
+		 * @name Collection#_array
+		 * @type {?Array}
+		 * @private
+		 */
+		Object.defineProperty(this, '_array', { value: null, writable: true, configurable: true });
+
+		/**
+		 * Cached array for the `keyArray()` method - will be reset to `null` whenever `set()` or `delete()` are called
+		 * @name Collection#_keyArray
+		 * @type {?Array}
+		 * @private
+		 */
+		Object.defineProperty(this, '_keyArray', { value: null, writable: true, configurable: true });
+	}
+
+	public set(key: K, value: V): this {
+		this._array = null;
+		this._keyArray = null;
+		return super.set(key, value);
+	}
+
+	public delete(key: K): boolean {
+		this._array = null;
+		this._keyArray = null;
+		return super.delete(key);
+	}
+
+	/**
+	 * Creates an ordered array of the values of this collection, and caches it internally. The array will only be
+	 * reconstructed if an item is added to or removed from the collection, or if you change the length of the array
+	 * itself. If you don't want this caching behavior, use `[...collection.values()]` or
+	 * `Array.from(collection.values())` instead.
+	 * @returns {Array}
+	 */
+	public array(): V[] {
+		if (!this._array || this._array.length !== this.size) this._array = [...this.values()];
+		return this._array;
+	}
+
+	/**
+	 * Creates an ordered array of the keys of this collection, and caches it internally. The array will only be
+	 * reconstructed if an item is added to or removed from the collection, or if you change the length of the array
+	 * itself. If you don't want this caching behavior, use `[...collection.keys()]` or
+	 * `Array.from(collection.keys())` instead.
+	 * @returns {Array}
+	 */
+	public keyArray(): K[] {
+		if (!this._keyArray || this._keyArray.length !== this.size) this._keyArray = [...this.keys()];
+		return this._keyArray;
+	}
+
+	/**
+	 * Obtains the first value(s) in this collection.
+	 * @param {number} [amount] Amount of values to obtain from the beginning
+	 * @returns {*|Array<*>} A single value if no amount is provided or an array of values, starting from the end if
+	 * amount is negative
+	 */
+	public first(): V | undefined;
+	public first(amount: number): V[];
+	public first(amount?: number): V | V[] | undefined {
+		if (typeof amount === 'undefined') return this.values().next().value;
+		if (amount < 0) return this.last(amount * -1);
+		amount = Math.min(this.size, amount);
+		const arr: V[] = Array.from({ length: amount });
+		const iter = this.values();
+		for (let i = 0; i < amount; i++) arr[i] = iter.next().value;
```

Comment:
Just a small nit, the [2nd param](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from#Parameters) within the Array.from() could be used here (and in some other places) as a mapping function.
```js
const iter = this.values();
return Array.from({ length: amount }, (): V => iter.next().value);
````

Reply:
Can also be applied to `firstKey`.